### ISSUE TITLE
ENH: Allow loading of subtree only, avoid crashes out of memory

### ISSUE THREAD

**Issue by corneliusroemer** [theosanderson/taxonium#372]
When browsing the full Usher tree, I often get crashes with "out of memory" despite having 32GB RAM.

It'd be great if one could specify a subtree to be loaded, e.g. using a query: load only the subtree with BA.2.75 or something like this - to reduce max memory usage.

Maybe this is done best in a matUtils command though. After all you're just showing trees.

**Comment by theosanderson** [comment]
Yep afraid this isn't planned and would be one for matUtils before generation. I think I can load the public tree with 8GB of RAM so those crashes aren't particularly expected. It may be about the amount of metadata.

**Comment by theosanderson** [comment]
To confirm: I can load the public tree with 5.85M sequences ( https://taxonium.org/?protoUrl=https://cov2tree.nyc3.cdn.digitaloceanspaces.com/latest_public.jsonl.gz ) in Chrome on my 8GB memory Macbook Pro. I can also load the full GISAID tree on my desktop, and while that has quite a bit of RAM, Chrome only reports using a few GB.

 It would be useful to know:

- when you experience crashes (whether it's a specific search op or something, or just loading)
- what browser your are using

Thanks!

**Comment by theosanderson** [comment]
Copying @corneliusroemer memory-related posts from  https://github.com/theosanderson/taxonium/issues/384
>Now tried on Chrome, there it crashed much sooner, entered a query after tree loaded and boom - gone.

>Error 5 - whatever that means

> Chrome tells me: paused before potential out of memory crash in filtering. That could be it? 
> ![image](https://user-images.githubusercontent.com/19732295/191061974-b269d9f1-56a8-4e4c-8972-7ce4283927e4.png)

> Could be on queries that return a lot of results - when I make a small query this doesn't happen.

**Comment by theosanderson** [comment]
Right: a small bit of process understanding this.

I've been able to reproduce these crashes on my Mac.

To summarise the situation:

- I can load the public tree (5M nodes) in Chrome on my MacBook Pro with 8 GB RAM
- I can load the full GISAID tree (>10M nodes) in Chrome on my Windows Desktop with >32 GB RAM

I cannot load the full GISAID tree in Chrome on my 8 GB Macbook Pro. Previously I had made sense of that by thinking that this tree required > 8 GB RAM. However I have just done some checking in DevTools and found that it maxes out at 2 GB of RAM usage in the web worker. So this is about that 2 GB limit. I'm now trying to understand that limit. I think the answer may be that on these platforms users will need to run their own Taxonium backends, which I can make more user friendly. Thanks for raising this.

**Comment by theosanderson** [comment]
OK so having done some tests, Firefox looks like the best solution for very large trees in MacOS. Previously I had warned against Firefox, as it loads initially quite a bit slower than Chrome (and also there were some issues in https://github.com/theosanderson/taxonium/issues/236 which later turned out to be unconnected).

However Firefox
- doesn't have this 2 GB limit on MacOS
- uses much less memory than Safari for the same sized tree

I have now:
- removed the warning not to use Firefox https://github.com/theosanderson/taxonium/commit/732e8caec78e8e7871cfd5ebb7fd4ba61483cb5e (though Firefox will still perform worse for non-huge trees)
- added a warning suggesting the use of Firefox when a user attempts to load a large tree in MacOS on Chrome #386. This could be made slicker.

I have added [a page](https://docs.taxonium.org/en/latest/loading_very_large_trees.html) to the documentation discussing these issues, but the description of how to use Docker to run a backend needs a bit of continued work.

**Comment by corneliusroemer** [comment]
Would it be possible to have a light version and make it heavy only for a
subtree?
I guess the issue is parsing the whole jsonl? So that may not help much :/

Unless there's some of index that you can use to read only parts of the
file into memory?

**Comment by corneliusroemer** [comment]
Firefox doesn't work for me with a 13m node tree. It freezes at this point:
<img width="1728" alt="image" src="https://user-images.githubusercontent.com/25161793/192050409-387c8c44-f80e-4f0e-9b3e-eb3f4acfe711.png">

**Comment by theosanderson** [comment]
Thanks for reporting. I have a solution coming very soon to allow users who want to interact with huge trees to easily start their own backend, more tonight or tomorrow.

**Comment by theosanderson** [comment]
OK, so suggested approach for huge trees is the following. (Still in beta). It requires Docker to be installed.

```
pip install -U taxoniumtools
view_taxonium my_great_tree.jsonl.gz
```

This should launch a local server that lets you explore the tree without issues.

**Comment by AngieHinrichs** [comment]
I'm giving this a try and so far so good.  The search seems slower, but (so far) not so much slower that I want to start over and take my chances with the crashes.  :)  Thanks Theo!

(FWIW FireFox seemed like it was going to take a very, very long time to load the full tree so I returned to Chrome, and when I checked on FireFox hours later it had quit with an out-of-mem error.)

**Comment by theosanderson** [comment]
Great! Thanks both for reporting the experience. Interesting and not particularly expected that search seems slower. Something to look at.

**Comment by corneliusroemer** [comment]
Just trying out taxoniumtools now, let's get them on bioconda if this works well!

**Comment by theosanderson** [comment]
Interested to hear how it goes!

For interest can you explain(/link to) the advantages that being on Bioconda has over just being on PyPI (and therefore still available in conda)?

https://github.com/theosanderson/taxonium/issues/393 would be a nice solution to this memory issue I think, but is a bit more work

**Comment by corneliusroemer** [comment]
Electron sounds interesting! Bioconda has advantage that one can see all the packages installed immediately, it can reduce storage requirements by recycling packages with identical versions between environments (if I'm not mistaken). If something is a pure python package, getting it into Bioconda and autoupdate is super easy.

I haven't really questioned putting things into bioconda (in addition), there seems not much of a downside AFAIK.
